modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java

```diff
@@ -343,35 +343,17 @@ void run(ClusterState state) {
                 continue;
             }
 
-            /*
-             * This is the pre-rollover write index. It may or may not be the write index after maybeExecuteRollover has executed, depending
-             * on rollover criteria. We're keeping a reference to it because regardless of whether it's rolled over or not we want to
-             * exclude it from force merging later in this data stream lifecycle run.
-             */
-            Index currentRunWriteIndex = dataStream.getWriteIndex();
-            try {
-                maybeExecuteRollover(state, dataStream);
-            } catch (Exception e) {
-                logger.error(
-                    () -> String.format(Locale.ROOT, "Data stream lifecycle failed to rollover data stream [%s]", dataStream.getName()),
-                    e
-                );
-                DataStream latestDataStream = clusterService.state().metadata().dataStreams().get(dataStream.getName());
-                if (latestDataStream != null) {
-                    if (latestDataStream.getWriteIndex().getName().equals(currentRunWriteIndex.getName())) {
-                        // data stream has not been rolled over in the meantime so record the error against the write index we
-                        // attempted the rollover
-                        errorStore.recordError(currentRunWriteIndex.getName(), e);
-                    }
-                }
-            }
-
+            // the following indices should not be considered for the remainder of this service run, for various reasons.
             Set<Index> indicesToExcludeForRemainingRun = new HashSet<>();
-            // the following indices should not be considered for the remainder of this service run:
-            // 1) the write index as it's still getting writes and we'll have to roll it over when the conditions are met
-            // 2) tsds indices that are still within their time bounds (i.e. now < time_series.end_time) - we don't want these indices to be
+
+            // This is the pre-rollover write index. It may or may not be the write index after maybeExecuteRollover has executed,
+            // depending on rollover criteria.
+            // The write index is excluded from the remaining run, as it's still getting writes and we'll have to roll it over
+            // when the conditions are met.
+            indicesToExcludeForRemainingRun.addAll(maybeExecuteRollover(state, dataStream));
```

Comment:
Changing `maybeExecuteRollover` in this way allows us to return both the backing write index and the failure store write index and exclude them from the next steps.